ddcommon-ffi/src/array_queue.rs

```diff
@@ -0,0 +1,296 @@
+    ) -> Self {
+        let internal_queue: crossbeam_queue::ArrayQueue<*mut c_void> =
+            crossbeam_queue::ArrayQueue::new(capacity);
+        let inner = Box::into_raw(Box::new(internal_queue)) as *mut c_void;
+        Self {
+            inner,
+            item_delete_fn,
+        }
+    }
+
+    pub fn take(&mut self) -> Option<Box<crossbeam_queue::ArrayQueue<*mut c_void>>> {
+        let raw = std::mem::replace(&mut self.inner, std::ptr::null_mut());
+
+        if raw.is_null() {
+            None
+        } else {
+            // # Safety: the raw pointer is not null and points to a valid memory location.
+            let queue =
+                unsafe { Box::from_raw(raw as *mut crossbeam_queue::ArrayQueue<*mut c_void>) };
+            while let Some(item) = queue.pop() {
+                // # Safety: the item is a valid memory location that can be deallocated by the
+                // item_delete_fn.
+                unsafe {
+                    (self.item_delete_fn)(item);
+                }
+            }
+            Some(queue)
+        }
+    }
+}
+
+impl Drop for ArrayQueue {
+    fn drop(&mut self) {
+        drop(self.take())
+    }
+}
+
+#[allow(unused)]
+#[repr(C)]
+pub enum ArrayQueueNewResult {
+    Ok(ArrayQueue),
+    Err(Error),
+}
+
+/// Creates a new ArrayQueue with the given capacity and item_delete_fn.
+/// The item_delete_fn is called when an item is dropped from the queue.
+#[no_mangle]
+#[must_use]
+pub extern "C" fn ddog_array_queue_new(
+    capacity: usize,
+    item_delete_fn: unsafe extern "C" fn(*mut c_void) -> c_void,
+) -> ArrayQueueNewResult {
+    let ffi_queue = ArrayQueue::new(capacity, item_delete_fn);
+    ArrayQueueNewResult::Ok(ffi_queue)
+}
+
+/// Converts a *mut ArrayQueue to a &mut crossbeam_queue::ArrayQueue<*mut c_void>.
+/// # Safety
+/// The pointer is null or points to a valid memory location allocated by array_queue_new.
+unsafe fn ddog_array_queue_ptr_to_inner<'a>(
+    queue_ptr: *mut ArrayQueue,
+) -> anyhow::Result<&'a mut crossbeam_queue::ArrayQueue<*mut c_void>> {
+    match queue_ptr.as_mut() {
+        None => anyhow::bail!("queue_ptr is null"),
+        Some(queue) => match queue.inner.as_mut() {
+            None => anyhow::bail!("queue.inner is null"),
+            Some(inner) => {
+                // # Safety: the inner points to a valid memory location which is a
+                // crossbeam_queue::ArrayQueue<*mut c_void>.
+                Ok(&mut *(inner as *mut c_void as *mut crossbeam_queue::ArrayQueue<*mut c_void>))
+            }
+        },
+    }
+}
+
+/// Drops the ArrayQueue.
+/// # Safety
+/// The pointer is null or points to a valid memory location allocated by array_queue_new.
+#[no_mangle]
+pub unsafe extern "C" fn ddog_array_queue_drop(queue_ptr: *mut ArrayQueue) {
+    if !queue_ptr.is_null() {
+        drop((*queue_ptr).take());
+    }
+}
+
+#[allow(unused)]
+#[repr(C)]
+pub enum ArrayQueuePushResult {
+    Ok,
+    Full(*mut c_void),
```

Comment:
Minor: It took me a while to understand why it was useful to return the value back, since for `push` it's somewhat redundant. Maybe include here a comment saying returning a value is mostly for `force_push`?